### PR TITLE
utils/ruby.sh: reoganize code

### PR DESCRIPTION
**Pull request by maxim-belkin**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Moved OS-specific logic to functions:
 - `find_ruby`: returns (via echo) Ruby that can be used
 - `usable_ruby`: checks whether Ruby satisfies Homebrew

Pre-create OS-specific error messages:
 - `upgrade_fail`: message that is printed when Homebrew fails to upgrade vendored Ruby
 - `install_fail`: message that is printed when Homebrew fails to install vendored Ruby

Unconditionally set `TERMINFO_DIRS` when installing/upgrading vendor Ruby **but** export it on Linux only.

**Motivation**: move OS-specific logic to functions so that it's easier to understand the logic in `setup-ruby-path`.

### CODE REVIEW

File: Library/Homebrew/utils/ruby.sh [hunk 1]
```diff
@@ -27,6 +55,16 @@ If there's no Homebrew Portable Ruby available for your processor:
...
 - try again
 "
+  local upgrade_fail
+  local install_fail
+  if [[ -n $HOMEBREW_MACOS ]]
+  then
+    upgrade_fail="Failed to upgrade Homebrew Portable Ruby!"
+    install_fail="Failed to install Homebrew Portable Ruby (and your system version is too old)!"
+  else
+    upgrade_fail="Failed to upgrade Homebrew Portable Ruby!$advice"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/ruby.sh:65]: How about defining `advice` in here too (as it's not used on macOS)?

File: Library/Homebrew/utils/ruby.sh [hunk 2]
```diff
@@ -37,68 +75,27 @@ If there's no Homebrew Portable Ruby available for your processor:
 
   unset HOMEBREW_RUBY_PATH
 
-  if [[ "$HOMEBREW_COMMAND" == "vendor-install" ]]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/ruby.sh:40]: I think this form is more explicit and easier to follow for people less familiar with Bash.

File: Library/Homebrew/utils/ruby.sh [hunk 3]
```diff
@@ -37,68 +75,27 @@ If there's no Homebrew Portable Ruby available for your processor:
...
-          HOMEBREW_RUBY_PATH=$ruby_exec
-          break
-        fi
-      done
-      IFS=$' \t\n' # Restore IFS to its default value
-    fi
+    HOMEBREW_RUBY_PATH=$(find_ruby)
 
-    if [[ -n "$HOMEBREW_MACOS_SYSTEM_RUBY_NEW_ENOUGH" ]]
+    if [[ -z "$HOMEBREW_RUBY_PATH" || -n "$HOMEBREW_FORCE_VENDOR_RUBY" ]] || ! usable_ruby
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/ruby.sh:91]: How about making this `unusable_ruby`?

**Comment by MikeMcQuaid** [comment]
Would this be to ease calling this from `cleanup.rb`?

**Comment by maxim-belkin** [comment]
> Would this be to ease calling this from `cleanup.rb`?

This is mainly to move work-performing (OS-specific) steps out from the `setup-ruby-path()` that manages the workflow. All those OS-dependent error messages and variable assignments were obfuscating the simple `if ... else...` workflow in `setup-ruby-path`.

Thanks for approving this!

**Pull request merged**